### PR TITLE
React to hosting changes

### PR DESCRIPTION
**Pull request by HaoK**
Putting back a precomp fallback SP for now since Hosting.Create is gone, UseServices is dead, so switched over to ConfigureServices everywhere

cc @davidfowl @rynowak @dougbu

### CODE REVIEW

File: src/Microsoft.AspNet.Mvc/RazorPreCompileModule.cs [hunk 1]
```diff
@@ -68,10 +58,12 @@ public virtual void BeforeCompile(IBeforeCompileContext context)
...
                 context.ProjectContext.ProjectDirectory);
-            serviceCollection.AddInstance<IApplicationEnvironment>(precompilationApplicationEnvironment);
 
-            var serviceProvider = serviceCollection.BuildServiceProvider();
-            var viewCompiler = new RazorPreCompiler(serviceProvider, context, _memoryCache, compilationSettings)
+            var replacedServices = new ServiceCollection();
+            replacedServices.AddMvc();
+            replacedServices.AddInstance<IApplicationEnvironment>(precompilationApplicationEnvironment);
+
+            var viewCompiler = new RazorPreCompiler(new WrappingServiceProvider(_appServices, replacedServices), context, _memoryCache, compilationSettings)
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc/RazorPreCompileModule.cs]: wrap long line

File: samples/MvcSample.Web/Startup.cs [hunk 2]
```diff
@@ -25,6 +25,65 @@ namespace MvcSample.Web
...
+            services.ConfigureMvc(options =>
+            {
+                options.Filters.Add(typeof(PassThroughAttribute), order: 17);
+                options.AddXmlDataContractSerializerFormatter();
+                options.Filters.Add(new FormatFilterAttribute());
+            });
+
+#if DNX451
+            //var applicationEnvironment = app.ApplicationServices.GetRequiredService<IApplicationEnvironment>();
+            //var configurationPath = Path.Combine(applicationEnvironment.ApplicationBasePath, "config.json");
```
Comment (dougbu) [inline samples/MvcSample.Web/Startup.cs]: - this code was necessary to get the functional tests working reliably when they invoked this web site.  why is it no longer necessary here?
- if fully-qualifying the configuration path isn't needed here, why did you leave it in `Configure(IApplicationBuilder)`?
- do not check in commented-out code

File: src/Microsoft.AspNet.Mvc/RazorPreCompileModule.cs (hunk 1 above)
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc/RazorPreCompileModule.cs]: or use a `var` for the `WrappingServiceProvider`

File: src/Microsoft.AspNet.Mvc/RazorPreCompileModule.cs [hunk 3]
```diff
@@ -84,6 +76,27 @@ public void AfterCompile(IAfterCompileContext context)
         {
         }
 
+        // REVIEW: Welcome back to life delegating SP!!!
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc/RazorPreCompileModule.cs]: ick!  we just got rid of this hack.  why doesn't Hosting support the approaches MVC requires?

File: src/Microsoft.AspNet.Mvc/RazorPreCompileModule.cs [hunk 4]
```diff
@@ -84,6 +76,27 @@ public void AfterCompile(IAfterCompileContext context)
...
+        private class WrappingServiceProvider : IServiceProvider
+        {
+            private readonly IServiceProvider _inner;
+            private readonly IServiceProvider _outer;
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc/RazorPreCompileModule.cs]: I have no idea what these names mean without reading the code.  use something like `_override` and `_fallback`

File: test/WebSites/ActionConstraintsWebSite/Startup.cs [hunk 5]
```diff
@@ -9,18 +9,19 @@ namespace ActionConstraintsWebSite
...
-                {
-                    options.AddXmlDataContractSerializerFormatter();
-                });
+                options.AddXmlDataContractSerializerFormatter();
             });
+        }
+
+        public void Configure(IApplicationBuilder app)
+        {
+            var configuration = app.GetTestConfiguration();
```
Comment (dougbu) [inline test/WebSites/ActionConstraintsWebSite/Startup.cs]: nit: bet you don't need the `configuration` variable anywhere

File: test/WebSites/AutofacWebSite/Startup.cs [hunk 6]
```diff
@@ -11,23 +11,24 @@ namespace AutofacWebSite
...
+            services.AddMvc();
+            services.AddTransient<HelloWorldBuilder>();
 
-            app.UseServices(services =>
-            {
-                services.AddMvc();
-                services.AddTransient<HelloWorldBuilder>();
+            var builder = new ContainerBuilder();
+            AutofacRegistration.Populate(builder,
+                                         services);
```
Comment (dougbu) [inline test/WebSites/AutofacWebSite/Startup.cs]: nit: please unwrap this line; very short

File: test/WebSites/ControllersFromServicesWebSite/Startup.cs [hunk 7]
```diff
@@ -16,34 +16,37 @@ namespace ControllersFromServicesWebSite
...
-                        .WithControllersAsServices(
-                         new[]
-                         {
+            services.AddMvc()
+                    .WithControllersAsServices(
+                     new[]
+                     {
                             typeof(TimeScheduleController).GetTypeInfo().Assembly
-                         });
+                     });
```
Comment (dougbu) [inline test/WebSites/ControllersFromServicesWebSite/Startup.cs]: fyi no need to line up the periods; just hit enter after `services` or `services.AddMvc()` and VS should do the Right Thing:tm:

File: test/WebSites/RazorCompilerCacheWebSite/Startup.cs [hunk 8]
```diff
@@ -9,17 +9,19 @@ namespace RazorCompilerCacheWebSite
 {
     public class Startup
     {
+        // Set up application services
+        public void ConfigureServices(IServiceCollection services)
+        {
+            // Add MVC services to the services container
```
Comment (dougbu) [inline test/WebSites/RazorCompilerCacheWebSite/Startup.cs]: nit: this comment is getting a little old by this point.  says nothing `services.AddMvc()` doesn't

File: test/WebSites/RazorCompilerCacheWebSite/Startup.cs [hunk 9]
```diff
@@ -9,17 +9,19 @@ namespace RazorCompilerCacheWebSite
 {
     public class Startup
     {
+        // Set up application services
```
Comment (dougbu) [inline test/WebSites/RazorCompilerCacheWebSite/Startup.cs]: isn't this actually setting up _request_ services?

File: test/WebSites/RoutingWebSite/Startup.cs [hunk 10]
```diff
@@ -8,18 +8,19 @@ namespace RoutingWebSite
...
+        }
 
-                services.AddScoped<TestResponseGenerator>();
-            });
+        public void Configure(IApplicationBuilder app)
+        {
+            var configuration = app.GetTestConfiguration();
 
-            app.UseErrorReporter();
+           app.UseErrorReporter();
```
Comment (dougbu) [inline test/WebSites/RoutingWebSite/Startup.cs]: nit: accidental change here

File: src/Microsoft.AspNet.Mvc/RazorPreCompileModule.cs (hunk 3 above)
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc/RazorPreCompileModule.cs]: put another way: we were using `HostingServices.Create()` and it was less evil / ugly than this

File: src/Microsoft.AspNet.Mvc/RazorPreCompileModule.cs [hunk 11]
```diff
@@ -84,6 +76,27 @@ public void AfterCompile(IAfterCompileContext context)
...
+                _outer = replacedServices.BuildServiceProvider();
+            }
+
+            public object GetService(Type serviceType)
+            {
+                return _outer.GetService(serviceType) ?? _inner.GetService(serviceType);
+            }
+        }
+
+
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc/RazorPreCompileModule.cs]: nit: remove extra blank line

**Comment by dougbu** [comment]
:watch:

File: test/WebSites/ActionConstraintsWebSite/Startup.cs (hunk 5 above)
Comment (HaoK) [inline test/WebSites/ActionConstraintsWebSite/Startup.cs]: I don't but it does magic I thought

File: test/WebSites/RazorCompilerCacheWebSite/Startup.cs (hunk 9 above)
Comment (HaoK) [inline test/WebSites/RazorCompilerCacheWebSite/Startup.cs]: no distinction anymore with hosting changes

File: test/WebSites/ActionConstraintsWebSite/Startup.cs (hunk 5 above)
Comment (dougbu) [inline test/WebSites/ActionConstraintsWebSite/Startup.cs]: I meant the line could just be `app.GetTestConfiguration();` (the method does magic; the variable does nothing)

Comment (HaoK) [inline test/WebSites/ActionConstraintsWebSite/Startup.cs]: I don't disagree but that seems like something that should be done as a full pass over the test code, personally I think you should rename the method

Comment (dougbu) [inline test/WebSites/ActionConstraintsWebSite/Startup.cs]: triage rejected the rename bug you had me file.  but you're moving an unused variable around and I wasn't suggesting additional changes to lines you weren't already touching.

Comment (HaoK) [inline test/WebSites/ActionConstraintsWebSite/Startup.cs]: My point is that this is an issue all over the test code base, I don't see the value in fixing one place, all that's going to do is make any future FindReplace All operations harder.

If this is something you think is worth fixing, I think it should be done as one PR doing a global replace all removing the var from all tests at once.

**Comment by HaoK** [comment]
Updated with changes, 7 test failures remaining
- Mostly TagHelpers, and RazorCompilerCacheTest.CompilerCache_IsNotInitializedUntilFirstViewRequest

File: test/Microsoft.AspNet.Mvc.FunctionalTests/ActionResultTests.cs [hunk 12]
```diff
@@ -3,6 +3,7 @@
 
 using System;
 using System.Collections.Generic;
+using System.Diagnostics;
```
Comment (davidfowl) [inline test/Microsoft.AspNet.Mvc.FunctionalTests/ActionResultTests.cs]: left over from Debugger.Launch

File: src/Microsoft.AspNet.Mvc/RazorPreCompileModule.cs (hunk 3 above)
Comment (davidfowl) [inline src/Microsoft.AspNet.Mvc/RazorPreCompileModule.cs]: You don't actually need this services to do precompilation.

Comment (davidfowl) [inline src/Microsoft.AspNet.Mvc/RazorPreCompileModule.cs]: I filed a bug for this. The RazorPreCompiler not use ISP in the ctor.

**Comment by HaoK** [comment]
Tests are all passing now

File: samples/MvcSample.Web/Startup.cs [hunk 13]
```diff
@@ -16,15 +16,75 @@
...
+
+            services.ConfigureMvc(options =>
+            {
+                options.Filters.Add(typeof(PassThroughAttribute), order: 17);
+                options.AddXmlDataContractSerializerFormatter();
+                options.Filters.Add(new FormatFilterAttribute());
+            });
+
+#if DNX451
+            // REVIEW: This is hella hacky
```
Comment (dougbu) [inline samples/MvcSample.Web/Startup.cs]: - replace this comment with a copy of the one in `Configure(IApplicationBuilder app)`
- this is not at all hacky compared to the previous approach of overriding the `static IServiceProvider` in all functional tests
- must remain as is until aspnet/Configuration#159 is fixed

File: samples/MvcSample.Web/Startup.cs [hunk 14]
```diff
@@ -16,15 +16,75 @@
...
+            var applicationEnvironment = services.BuildServiceProvider().GetRequiredService<IApplicationEnvironment>();
+            var configurationPath = Path.Combine(applicationEnvironment.ApplicationBasePath, "config.json");
+
+            // Set up configuration sources.
+            var configuration = new Configuration()
+                .AddJsonFile(configurationPath)
+                .AddEnvironmentVariables();
+            string diSystem;
+            if (configuration.TryGet("DependencyInjection", out diSystem) &&
+                diSystem.Equals("AutoFac", StringComparison.OrdinalIgnoreCase))
```
Comment (dougbu) [inline samples/MvcSample.Web/Startup.cs]: can the code from line 48 to here be done once?  that is, does Hosting create a single `Startup` instance and therefore support use of a `bool _isAutoFac;` field?

File: samples/MvcSample.Web/Startup.cs [hunk 15]
```diff
@@ -16,15 +16,75 @@
...
+            {
+
```
Comment (dougbu) [inline samples/MvcSample.Web/Startup.cs]: nit: remove extra blank line

File: src/Microsoft.AspNet.Mvc.TagHelpers/ScriptTagHelper.cs [hunk 16]
```diff
@@ -3,6 +3,7 @@
 
 using System;
 using System.Collections.Generic;
+using System.Diagnostics;
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.TagHelpers/ScriptTagHelper.cs]: ??

File: samples/MvcSample.Web/Startup.cs [hunk 17]
```diff
@@ -16,15 +16,75 @@
 #if DNX451
 using Microsoft.Framework.DependencyInjection.Autofac;
 using Microsoft.Framework.Runtime;
+using Microsoft.Framework.Runtime.Infrastructure;
 #endif
+using System.ComponentModel;
```
Comment (dougbu) [inline samples/MvcSample.Web/Startup.cs]: what brought this in?

File: src/Microsoft.AspNet.Mvc/RazorViewEngineOptionsSetup.cs [hunk 18]
```diff
@@ -1,6 +1,7 @@
 ﻿// Copyright (c) Microsoft Open Technologies, Inc. All rights reserved.
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
+using System.Diagnostics;
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc/RazorViewEngineOptionsSetup.cs]: ??

File: test/Microsoft.AspNet.Mvc.FunctionalTests/BestEffortLinkGenerationTest.cs [hunk 19]
```diff
@@ -15,6 +16,8 @@ public class BestEffortLinkGenerationTest
     {
         private const string SiteName = nameof(BestEffortLinkGenerationWebSite);
         private readonly Action<IApplicationBuilder> _app = new BestEffortLinkGenerationWebSite.Startup().Configure;
+        private readonly Action<IServiceCollection> _configureServices = new BestEffortLinkGenerationWebSite.Startup().ConfigureServices;
+
```
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.FunctionalTests/BestEffortLinkGenerationTest.cs]: nit: remove extra blank lines added after a number of these new fields

File: test/Microsoft.AspNet.Mvc.FunctionalTests/TestHelper.cs [hunk 20]
```diff
@@ -81,10 +113,11 @@ public static TestServer CreateServer(Action<IApplicationBuilder> builder, strin
                 applicationBasePath,
                 applicationWebSiteName);
             services.AddInstance<IApplicationEnvironment>(environment);
+            services.AddInstance<IHostingEnvironment>(new HostingEnvironment(environment));
```
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.FunctionalTests/TestHelper.cs]: from the way you've updated the functional tests, it appears this is the only required change to this class.  plus likely the `using` updates.  revert adding the new, unused `CreateServer`, the `private` method renaming, and the minor change below.

**Comment by dougbu** [comment]
:watch:

File: test/Microsoft.AspNet.Mvc.FunctionalTests/TestHelper.cs (hunk 20 above)
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.FunctionalTests/TestHelper.cs]: @HaoK please revert the changes to this file that were not required.  we don't need unused helper methods lying around.

File: src/Microsoft.AspNet.Mvc/RazorPreCompileModule.cs (hunk 3 above)
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc/RazorPreCompileModule.cs]: why did this "REVIEW" comment get checked in?  we don't do TODOs in the code

**Comment by dougbu** [comment]
4b5dd19
